### ISSUE TITLE
Key not present in Dictionary: "ocamlmerlin_path"

### ISSUE THREAD

**Issue by arobase-che** [ocaml/merlin#1408]
Hi,

I have some problem making merlin works within vim.

I simply follow the documentation.
Installing opam (`pacman -S opam darcs`). Them `opam init`.
Yes, I let opam modify the configuration of fish (my shell).
From here, there is a message saying that I should do `eval (opam env --switch=default)` so I do that.

Now, I simply install merlin. `opam install merlin`. Letting install everything it want to. (Y). 
From here, I have a message `Quick setup for VIM`. Asking me to add that :  

~~~plain
let g:opamshare = substitute(system('opam var share'),'\n$','','''')
execute "set rtp+=" . g:opamshare . "/merlin/vim"
~~~

To vim, so I do it.

I don't really wan to install `opam user-setup install` so before installing it, I just check `vim test.ml` and *I got* `pynvim.api.common.NvimError: Vim(return):E716: Key not present in Dictionary: "ocamlmerlin_path"` (yeah,  neovim is aliased on vim but I got the same error on the real vim). `opam user-setup install` does help. It just install itself saying `[WARNING] var was deprecated in version 2.1 of the opam CLI. Use opam var instead or set OPAMCLI environment variable to 2.0.`.

Here is `opam env` : 
~~~plain
set -gx OPAM_SWITCH_PREFIX '/home/ache/.opam/default';
set -gx CAML_LD_LIBRARY_PATH '/home/ache/.opam/default/lib/stublibs:/usr/lib/ocaml/stublibs:/usr/lib/ocaml';
set -gx OCAML_TOPLEVEL_PATH '/home/ache/.opam/default/lib/toplevel';
builtin -n | /bin/sh -c 'grep -q \'^argparse$\'' 1>/dev/null 2>/dev/null; and set -gx MANPATH ':/home/ache/.opam/default/man';
set -gx PATH '/home/ache/.opam/default/bin' '/usr/local/sbin' '/usr/local/bin' '/usr/bin' '/usr/bin/site_perl' '/usr/bin/vendor_perl' '/usr/bin/core_perl' '/home/ache/.npm-packages/bin' '/home/ache/.go/bin' '/home/ache/.cargo/bin' '/usr/lib/emscripten' '.' '/home/ache/.npm-packages/bin' '/home/ache/.go/bin' '/home/ache/.cargo/bin' '/usr/lib/emscripten' '.';
~~~


`opam switch`
~~~plain
#   switch   compiler      description
->  default  ocaml.4.12.0  default

[WARNING] The environment is not in sync with the current switch.
          You should run: eval (opam env)
~~~

It keeps saying that I should run `eval (opam env)` every time I launch my shell. But doing so doesn't solve my problem with merlin.

So what did I do wrong ? :( 
Did I miss something 



Here is the full error message from vim when I open a `.ml` file. I got a similar error trying to have completion of a word.

~~~plain
Error detected while processing function merlin#Register[147]..merlin#LoadProject[2]..provider#python3#Call:                                                                                                                                                                   
line   18:                                                                                                                                                                                                                                                                     
Error invoking 'python_execute' on channel 3 (python3-script-host):                                                                                                                                                                                                            
Traceback (most recent call last):                                                                                                                                                                                                                                             
  File "<string>", line 1, in <module>                                                                                                                                                                                                                                         
  File "/home/ache/.opam/default/share/merlin/vim/autoload/merlin.py", line 876, in setup_merlin                                                                                                                                                                               
    result = command("check-configuration")                                                                                                                                                                                                                                    
  File "/home/ache/.opam/default/share/merlin/vim/autoload/merlin.py", line 189, in command                                                                                                                                                                                    
    return command2(args)                                                                                                                                                                                                                                                      
  File "/home/ache/.opam/default/share/merlin/vim/autoload/merlin.py", line 173, in command2                                                                                                                                                                                   
    result = json.loads(merlin_exec(cmdline,input=content))                                                                                                                                                                                                                    
  File "/home/ache/.opam/default/share/merlin/vim/autoload/merlin.py", line 99, in merlin_exec                                                                                                                                                                                 
    path = vim.eval("merlin#SelectBinary()")                                                                                                                                                                                                                                   
  File "/usr/lib/python3.9/site-packages/pynvim/plugin/script_host.py", line 205, in eval                                                                                                                                                                                      
    obj = self.request("vim_eval", expr)                                                                                                                                                                                                                                       
  File "/usr/lib/python3.9/site-packages/pynvim/api/nvim.py", line 182, in request                                                                                                                                                                                             
    res = self._session.request(name, *args, **kwargs)                                                                                                                                                                                                                         
  File "/usr/lib/python3.9/site-packages/pynvim/msgpack_rpc/session.py", line 102, in request                                                                                                                                                                                  
    raise self.error_wrapper(err)                                                                                                                                                                                                                                              
pynvim.api.common.NvimError: Vim(return):E716: Key not present in Dictionary: "ocamlmerlin_path"      
~~~

I have `ocamlmerlin` in my path.

Thank you !

**Comment by arobase-che** [comment]
I just retry the exact same process in a LXC container. Using debian with bash and vim.nox. And I have the exact same problem : 

~~~plain
vim.error: Vim(return):E716: Key not present in Dictionary: "ocamlmerlin_path"
~~~

I guess I should have a look further but if this repository isn't broken then I must have miss something. :/

**Comment by voodoos** [comment]
I just tried in a fresh ubuntu vm.
- Opam installed from the package manager.
- `opam install merlin`
- This `.vimrc`:
```
let g:opamshare = substitute(system('opam var share'),'\n$','','''')
execute "set rtp+=" . g:opamshare . "/merlin/vim"
syntax on
filetype plugin on
filetype indent on
```

And everything works as expected...

**Comment by arobase-che** [comment]
Ok, so I checked this morning (2 weeks delay sorry ...).

I started a Ubuntu container and as you said work like a charm.

So I checked a lot a thinks and finally found the problem.
I should add : 
~~~vim
let g:opamshare = substitute(system('opam var share'),'\n$','','''')
execute "set rtp+=" . g:opamshare . "/merlin/vim"
~~~

To `init.vim` because it had to be run early. And I didn't add it here but to an other configuration file, so it was loaded too late.

Thank you for taking time to launch a VM, It give me access to a functional setup that I could compare to mine.
